### ISSUE TITLE
Label the TLS in vmmap

### ISSUE THREAD

**Issue by gsingh93** [pwndbg/pwndbg#1570]
The `tls` command can find the TLS address, so we know where it is, but this doesn't show up in `vmmap`:
```
pwndbg> tls
Thread Local Storage (TLS) base: 0x7ffff7cb3800
pwndbg> vmmap
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
             Start                End Perm     Size Offset File
    0x555555554000     0x555555558000 r--p     4000      0 /usr/bin/ls
    0x555555558000     0x55555556c000 r-xp    14000   4000 /usr/bin/ls
    0x55555556c000     0x555555574000 r--p     8000  18000 /usr/bin/ls
    0x555555575000     0x555555576000 r--p     1000  20000 /usr/bin/ls
    0x555555576000     0x555555577000 rw-p     1000  21000 /usr/bin/ls
    0x555555577000     0x555555599000 rw-p    22000      0 [heap]
    0x7ffff7cb3000     0x7ffff7cb6000 rw-p     3000      0 [anon_7ffff7cb3]
    ...
```

It would be useful to add this information into vmmap.

**Comment by disconnect3d** [comment]
Hm... if we do it, then any `vmmap` command or context display may call to `pthread_self()` via the `pwndbg.gdblib.tls.address`. Of course we can cache/memoize this but the question is if this is totally safe to do and does not trigger any side effects since we will be calling a function in the target process?

**Comment by gsingh93** [comment]
That's a good question, but if we're concerned about that then doesn't that mean the `tls` command itself is unsafe to call? Might not be a bad idea to just add this to `vmmap` and if anything breaks, we've found a bug in the `tls` command :)

**Comment by disconnect3d** [comment]
Sure, but the `tls` command or tls base address is needed less often than getting vmmaps itself.

Btw I usually fetched the TLS address by using the `canary` command and then searching for the global canary value, as, at least with glibc, it is stored in TLS:
<img width="643" alt="image" src="https://user-images.githubusercontent.com/10009354/219102291-9375cc54-30cb-4977-85be-a0841bf6be8b.png">

---

Regarding the `tls` command/fetching, here is what it does in full:
* On x86-64 it will try to use the `fsbase` register:
  * on x86-64 and GDB >= 8.0 (or 7.9?) it will return the `fsbase` register fetched from GDB
  * if its not remote debug session, it will return the `fsbase` register fetched with `ptrace` syscall
  * otherwise, returns 0
* On i386 it will try to use the `gsbase` register
* On aarch64 it will try to use the `TPIDR_EL0` register

If the fetched register value is equal to 0, or if we are on a different architecture, it will fallback to calling the `pthread_self()`.

Now, what this function really does is it "obtains ID of the calling thread". Why is that really the address of TLS? Is that true on all architectures, glibc versions or variations of libraries that are linked? I don't know, but I am afraid this call may not give us the correct result in all cases (cc: @lebr0nli).

Interestingly, on a Firefox process, it actually is the same:
```
pwndbg> tls
Thread Local Storage (TLS) base: 0x7fa19979d380
pwndbg> pi hex(pwndbg.gdblib.tls.call_pthread_self())
'0x7fa19979d380'
pwndbg> fsbase
0x7fa19979d380
```

So maybe it is always correct... No idea 🤐

**Comment by lebr0nli** [comment]
> I don't know, but I am afraid this call may not give us the correct result in all cases

You are right, glibc’s TLS implementation is architecture-specific, and the address it returns might not always be correct.
This is my bad, I didn’t notice that before.

If I didn’t misunderstand something:

The implements of `pthread_self()` is:
https://elixir.bootlin.com/glibc/glibc-2.37/source/nptl/pthread_self.c#L22
`THREAD_SELF` is defined as:
i386: https://elixir.bootlin.com/glibc/glibc-2.37/source/sysdeps/i386/nptl/tls.h#L234
x86-64: https://elixir.bootlin.com/glibc/glibc-2.37/source/sysdeps/x86_64/nptl/tls.h#L181
arm: https://elixir.bootlin.com/glibc/latest/source/sysdeps/arm/nptl/tls.h#L76

So the reason that `pthread_self()` can return the TLS base on i386 and x86-64 is that it returns the self-reference of the TLS:
https://elixir.bootlin.com/glibc/glibc-2.37/source/nptl/pthread_create.c#L671
(If we overwrite this value, `pthread_self()` will return the wrong value)

For arm, the implementation of `THREAD_SELF` has different from i386 and x86-64, we need to add back `sizeof(struct pthread)` to the result to get the correct base address of TLS.
And for the others architecture, the implementation might have different from the above of them.

----

To summarize, calling `pthread_self()` might help on some of the architectures, but it is not always correct, and it's probably not a good idea to call it implicitly.
I'm working on the fix for it now, I'll open a draft PR(#1579) soon for this and combine it with some of the enhancements for the issues about heap heuristics I mentioned in the Discord. (since I created these TLS functions mainly for the heap heuristics at the beginning)

**Comment by disconnect3d** [comment]
Btw what else is stored in that page? Is all of that TLS?

**Comment by Tomatotech90** [comment]
The TLS region appears as [anon_...] in vmmap output while [heap] and [stack] are already labeled. The fix would label the matching region as [tls].

**Comment by disconnect3d** [comment]
@Tomatotech90 Thanks for your interest! I deleted your post about working on this/assigning to the issue as we generally don't need to assign contributors to issues ahead of time. If you want to work on this, please feel free to investigate this and open a PR when you have the fix/feature. Also please see our [Contributing documentation](https://pwndbg.re/dev/contributing/) for more info on contributing to the project.


Regarding this:

> The TLS region appears as [anon_...] in vmmap output while [heap] and [stack] are already labeled. The fix would label the matching region as [tls].

This does not answer my question. If you look closely here:
```
pwndbg> tls
Thread Local Storage (TLS) base: 0x7ffff7cb3800
pwndbg> vmmap
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
             Start                End Perm     Size Offset File
    ...
    0x7ffff7cb3000     0x7ffff7cb6000 rw-p     3000      0 [anon_7ffff7cb3]
    ...
```
The page starts at 0x...3000 whereas the TLS base is 0x...3800 so there is at least 0x800 bytes that are not TLS. I bet the TLS also isn't as big as 0x3000-0x800 bytes, so the other part of the page may also not be TLS.

**Comment by Tomatotech90** [comment]
> [@Tomatotech90](https://github.com/Tomatotech90) Thanks for your interest! I deleted your post about working on this/assigning to the issue as we generally don't need to assign contributors to issues ahead of time. If you want to work on this, please feel free to investigate this and open a PR when you have the fix/feature. Also please see our [Contributing documentation](https://pwndbg.re/dev/contributing/) for more info on contributing to the project.
> 
> Regarding this:
> 
> > The TLS region appears as [anon_...] in vmmap output while [heap] and [stack] are already labeled. The fix would label the matching region as [tls].
> 
> This does not answer my question. If you look closely here:
> 
> ```
> pwndbg> tls
> Thread Local Storage (TLS) base: 0x7ffff7cb3800
> pwndbg> vmmap
> LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
>              Start                End Perm     Size Offset File
>     ...
>     0x7ffff7cb3000     0x7ffff7cb6000 rw-p     3000      0 [anon_7ffff7cb3]
>     ...
> ```
> 
> The page starts at 0x...3000 whereas the TLS base is 0x...3800 so there is at least 0x800 bytes that are not TLS. I bet the TLS also isn't as big as 0x3000-0x800 bytes, so the other part of the page may also not be TLS.

Thanks for redirecting me to the guidance and the feedback. The first comment only addressed the missing label that I thought was the issue. After inspecting the full page layout, it appears that the bytes before the TLS base are zero padding for alignment, followed by the struct pthread, then the DTV, all part of the TLS allocation. Before opening the PR, is labeling the entire page [tls] the right approach, given that the TLS base is not page aligned?

**Comment by k4lizen** [comment]
some notes

if this mapping is not only TLS, we could consider adding a `<--- TLS` annotation similarly to how we do for `vis`

need to care that it works properly cross-arch.

need to care and document the selinux gotcha we had

**Comment by Tomatotech90** [comment]
Thanks for the notes @k4lizen. The <--- TLS annotation approach makes sense given that the TLS base is not page-aligned and the page contains padding before it. I can revise the implementation to annotate the TLS base address in the vmmap output rather than relabeling the entire page, similar to how vis does.
